auth/credentials/oauth.go

```diff
@@ -66,142 +64,82 @@ func (c *OAuthTokenSource) RevokeToken() error {
+func ReuseWrapperCacheTokenSource(src oauth2.TokenSource, cache LocalTokenCache) oauth2.TokenSource {
+	return oauth2.ReuseTokenSource(nil, &tokenSourceWrapper{src: src, cache: cache})
+}
 
-	return c.token, nil
+type tokenSourceWrapper struct {
+	src   oauth2.TokenSource
+	cache LocalTokenCache
 }
 
-// refreshToken fetches a new Token and saves it using the Token source.
-func (c *OAuthTokenSource) refreshToken() (*Token, error) {
-	newToken, err := c.fetchTokenFromRemoteServer()
+func (w *tokenSourceWrapper) Token() (*oauth2.Token, error) {
+	tokenString, err := w.cache.RetrieveToken(context.Background())
+	if err == nil && tokenString != "" {
+		token, err := parseToken(tokenString)
+		if err == nil && !isTokenExpired(token) {
+			return &oauth2.Token{
+				AccessToken: token.AccessToken,
+				Expiry:      token.Expiry,
+			}, nil
+		}
+	}
+
+	token, err := w.src.Token()
 	if err != nil {
 		return nil, err
 	}
 
-	// Save the access Token string to the Token source
-	err = c.tokenCache.SaveToken(c.ctx, newToken.AccessToken)
+	err = w.cache.SaveToken(context.Background(), token.AccessToken)
 	if err != nil {
-		return nil, fmt.Errorf("failed to save Token: %w", err)
+		return nil, err
 	}
 
-	c.token = newToken
-	return newToken, nil
+	return token, nil
 }
 
 // ExpiryDelta is the Additional time for Access Tokens to not expire.
 const ExpiryDelta = 10 * time.Second
 
 // expired checks if the Token is close to expiring.
-func (t *Token) expired() bool {
-	if t.Expiry.IsZero() {
+func isTokenExpired(token *oauth2.Token) bool {
+	if token.Expiry.IsZero() {
 		return false
 	}
-	return t.Expiry.Round(0).Add(-ExpiryDelta).Before(time.Now())
+	return token.Expiry.Round(0).Add(-ExpiryDelta).Before(time.Now())
 }
 
-// Valid checks if the Token is still valid (present and not expired)
-func (t *Token) Valid() bool {
-	return t != nil && t.AccessToken != "" && !t.expired()
-}
-
-// ParseToken extracts expiry details from JWT Token
-func parseToken(accessToken string) (*Token, error) {
+// parseToken extracts expiry details from JWT Token
+func parseToken(accessToken string) (*oauth2.Token, error) {
 	parts := strings.Split(accessToken, ".")
 	if len(parts) != 3 {
 		return nil, errors.New("invalid access Token format")
 	}
-
 	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
 	if err != nil {
 		return nil, err
 	}
-
 	var tokenData struct {
 		Exp int64 `json:"exp"`
 	}
 	if err := json.Unmarshal(payload, &tokenData); err != nil {
 		return nil, err
 	}
-
 	expiry := time.Unix(tokenData.Exp, 0)
 	if time.Now().After(expiry) {
-		return nil, errors.New("Atlas Cloud Access Token has expired")
+		return nil, errors.New("atlas cloud Access Token has expired")
 	}
-
-	return &Token{
+	return &oauth2.Token{
 		AccessToken: accessToken,
 		Expiry:      expiry,
-		ExpiresIn:   int(time.Until(expiry).Seconds()),
 	}, nil
 }
 
-// fetchTokenFromRemoteServer makes a manual POST request to Server (tokenUrl) to fetch the access Token.
```

Comment:
This is currently done by oauth2 lib. 
With this change we're missing the ability to set User-Agent and control error handling, retry logic etc.